docker/hail-ubuntu/controller.sh

```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
+
+set -ex
+
+COMMAND=$*
+
+restart() {
+    kill $PID
+    sleep 1
+    if kill -0 $PID >/dev/null 2>&1
+    then
+        sleep 5
+        kill -9 $PID || true
+    fi
+
+    $COMMAND &
+    PID=$!
+    restarted=yes
+}
+trap restart SIGINT
+
+term() {
+    kill -TERM $PID
+}
+trap term SIGTERM
+
+
+$COMMAND &
+PID=$!
+restarted=yes
+
+while [ $restarted == "yes" ]
+do
+    restarted=no
```

Comment:
Can you describe in words how this all works together? I understand the traps and the code in `restart`. I just don't see how this loop works with `restarted`.